barretenberg/cpp/src/barretenberg/commitment_schemes/small_subgroup_ipa/small_subgroup_ipa.hpp

```diff
@@ -0,0 +1,415 @@
+    /*!
+     * @brief Verifies the consistency of polynomial evaluations provided by the prover.
+     *
+     * @details
+     * Given a subgroup of \f$ \mathbb{F}^\ast \f$, its generator \f$ g\f$, this function checks whether the following
+     * equation holds:
+     * \f[ L_1(r) A(r) + (r - g^{-1}) \left( A(g*r) - A(r) - F(r) G(r) \right) + L_{|H|}(r) \left( A(r) - s
+     * \right) = T(r) Z_H(r) \f] Where the following are sent by the prover
+     * - \f$ A(r), A(g\cdot r) \f$ are the evaluation of the "big sum polynomial"
+     * - \f$ G(r) \f$ is the evaluation of the concatenation of the coefficients of the masking Libra polynomials
+     *
+     * - \f$ T(r) \f$ is the evaluation of the quotient of the left hand side above by the vanishing polynomial for
+     * \f$H\f$
+     * and the following evaluations computed by the verifier
+     * - \f$ L_1 \f$ and \f$ L_{|H|} \f$ are the Lagrange polynomials corresponding to \f$ 1 \f$ and \f$ g^{-1} \f$.
+     * - \f$ F(r) \f$ is the evaluation of the polynomial obtained by concatenating powers of sumcheck round challenges
+     * - \f$ Z_H(r) \f$ is the vanishing polynomial \f$ X^{|H|} - 1\f$ evaluated at the challenge point.
+     *
+     * @param libra_evaluations A vector of polynomial evaluations containing:
+     *   - \f$ G(r), A(g\cdot r), A(r), T(r) \f$.
+     * @param gemini_evaluation_challenge The challenge point \f$ r \f$ at which evaluations are verified.
+     * @param multilinear_challenge A vector of sumcheck round challenges.
+     * @param eval_claim The claimed inner proudct of the coefficients of \f$G\f$ and \f$F\f$.
+     * @return True if the consistency check passes, false otherwise.
+     */
+    static bool evaluations_consistency_check(const std::vector<FF>& libra_evaluations,
+                                              const FF& gemini_evaluation_challenge,
+                                              const std::vector<FF>& multilinear_challenge,
+                                              const FF& inner_product_eval_claim)
+    {
+
+        static const FF subgroup_generator_inverse = FF(1) / Curve::SUBGROUP_GENERATOR;
+
+        // Compute the evaluation of the vanishing polynomia Z_H(X) at X = gemini_evaluation_challenge
+        const FF vanishing_poly_eval = gemini_evaluation_challenge.pow(SUBGROUP_SIZE) - FF(1);
+
+        // Construct the challenge polynomial from the sumcheck challenge, the verifier has to evaluate it on its own
+        const std::vector<FF> challenge_polynomial_lagrange = compute_challenge_polynomial(multilinear_challenge);
+
+        // Compute the evaluations of the challenge polynomial, Lagrange first, and Lagrange last for the fixed small
+        // subgroup
+        auto [challenge_poly, lagrange_first, lagrange_last] = compute_batched_barycentric_evaluations(
+            challenge_polynomial_lagrange, gemini_evaluation_challenge, subgroup_generator_inverse);
+
+        const FF& concatenated_at_r = libra_evaluations[0];
```

Comment:
Why isn't `libra_evaluations` a struct? If we only do this accessing once ok, otherwise consider refactor.

Reply:
It's done only during this check